Signal/src/view controllers/CallViewController.swift

```diff
@@ -541,27 +630,87 @@ class CallViewController: UIViewController, CallObserver {
     // MARK: - CallObserver
 
     internal func stateDidChange(call: SignalCall, state: CallState) {
-        DispatchQueue.main.async {
-            Logger.info("\(self.TAG) new call status: \(state)")
-            self.updateCallUI(callState: state)
-        }
+        AssertIsOnMainThread()
+        Logger.info("\(self.TAG) new call status: \(state)")
+        self.updateCallUI(callState: state)
     }
 
     internal func hasVideoDidChange(call: SignalCall, hasVideo: Bool) {
-        DispatchQueue.main.async {
-            self.updateCallUI(callState: call.state)
-        }
+        AssertIsOnMainThread()
+        self.updateCallUI(callState: call.state)
     }
 
     internal func muteDidChange(call: SignalCall, isMuted: Bool) {
-        DispatchQueue.main.async {
-            self.updateCallUI(callState: call.state)
-        }
+        AssertIsOnMainThread()
+        self.updateCallUI(callState: call.state)
     }
 
     internal func speakerphoneDidChange(call: SignalCall, isEnabled: Bool) {
-        DispatchQueue.main.async {
-            self.updateCallUI(callState: call.state)
+        AssertIsOnMainThread()
+        self.updateCallUI(callState: call.state)
+    }
+
+    // MARK: - Video
+
+    internal func updateLocalVideoTrack(localVideoTrack: RTCVideoTrack?) {
+        AssertIsOnMainThread()
+        if self.localVideoTrack == localVideoTrack {
+            return
         }
+
+        self.localVideoTrack = localVideoTrack
+
+        var source: RTCAVFoundationVideoSource?
+        if localVideoTrack?.source is RTCAVFoundationVideoSource {
+            source = localVideoTrack?.source as! RTCAVFoundationVideoSource
+        }
+        localVideoView.captureSession = source?.captureSession
+        let isHidden = source == nil
+        Logger.info("\(TAG) \(#function) isHidden: \(isHidden)")
+        localVideoView.isHidden = source == nil
+
+        updateVideoViewLayout()
+    }
+
+    internal func updateRemoteVideoTrack(remoteVideoTrack: RTCVideoTrack?) {
+        AssertIsOnMainThread()
+        if self.remoteVideoTrack == remoteVideoTrack {
+            return
+        }
+
+        self.remoteVideoTrack?.remove(remoteVideoView)
+        self.remoteVideoTrack = nil
+        remoteVideoView.renderFrame(nil)
```

Comment:
Right, that's my best guess too, but the docs aren't very forthcoming. Anyway, I'm doing this because the demo app does and it certainly makes sense that we'd want to flush the contents of the remote video view.